modules/optimization/src/executioners/OptimizeSolve.C

```diff
@@ -0,0 +1,409 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "OptimizeSolve.h"
+#include "OptimizationAppTypes.h"
+
+#include "libmesh/petsc_vector.h"
+#include "libmesh/petsc_matrix.h"
+
+InputParameters
+OptimizeSolve::validParams()
+{
+  InputParameters params = emptyInputParameters();
+  MooseEnum tao_solver_enum("taontr taobntr taobncg taonls taobnls taontl taobntl taolmvm "
+                            "taoblmvm taonm taobqnls taoowlqn taogpcg taobmrm");
+  params.addRequiredParam<MooseEnum>(
+      "tao_solver", tao_solver_enum, "Tao solver to use for optimization.");
+  ExecFlagEnum exec_enum = ExecFlagEnum();
+  exec_enum.addAvailableFlags(EXEC_NONE,
+                              OptimizationAppTypes::EXEC_FORWARD,
+                              OptimizationAppTypes::EXEC_ADJOINT,
+                              OptimizationAppTypes::EXEC_HOMOGENEOUS_FORWARD);
+  exec_enum = {OptimizationAppTypes::EXEC_FORWARD,
+               OptimizationAppTypes::EXEC_ADJOINT,
+               OptimizationAppTypes::EXEC_HOMOGENEOUS_FORWARD};
+  params.addParam<ExecFlagEnum>(
+      "solve_on", exec_enum, "List of flags indicating when inner system solve should occur.");
+  return params;
+}
+
+OptimizeSolve::OptimizeSolve(Executioner & ex)
+  : SolveObject(ex),
+    _my_comm(MPI_COMM_SELF),
+    _solve_on(getParam<ExecFlagEnum>("solve_on")),
+    _verbose(getParam<bool>("verbose")),
+    _tao_solver_enum(getParam<MooseEnum>("tao_solver").getEnum<TaoSolverEnum>()),
+    _parameters(std::make_unique<libMesh::PetscVector<Number>>(_my_comm))
+{
+  if (libMesh::n_threads() > 1)
+    mooseError("OptimizeSolve does not currently support threaded execution");
+}
+
+bool
+OptimizeSolve::solve()
+{
+  // Initial solve
```

Comment:
I think you may want a timed section here for the perf graph

Reply:
done